scio-core/src/main/scala/com/spotify/scio/io/FileStorage.scala

```diff
@@ -91,45 +103,44 @@ final private[scio] class FileStorage(protected[scio] val path: String) {
         case _: Throwable => buffered
       }
     }
-    val input = getDirectoryInputStream(path, wrapInputStream)
+    val input = getDirectoryInputStream(wrapInputStream)
     IOUtils.lineIterator(input, StandardCharsets.UTF_8).asScala
   }
 
-  def isDone: Boolean = {
-    val partPattern = "([0-9]{5})-of-([0-9]{5})".r
-    val metadata =
-      try {
-        listFiles
-      } catch {
-        case _: FileNotFoundException => Seq.empty
-      }
-    val nums = metadata.flatMap { meta =>
-      val m = partPattern.findAllIn(meta.resourceId().toString)
-      if (m.hasNext) {
-        Some((m.group(1).toInt, m.group(2).toInt))
-      } else {
-        None
+  def isDone(): Boolean = {
+    val files = Try(listFiles).recover { case _: FileNotFoundException => Seq.empty }.get
+
+    // best effort matching shardNumber and numShards
+    // relies on the shardNameTemplate to be of '$prefix-$shardNumber-of-$numShards$suffix' format
+    val writtenShards = files
+      .map(_.resourceId().toString)
+      .flatMap {
+        case ShardPattern(prefix, shardNumber, numShards, suffix) =>
+          val part = for {
+            idx <- Try(shardNumber.toInt)
+            total <- Try(numShards.toInt)
+            // prefix or suffix may contain pane/window info
+            key = (prefix, suffix, total)
+          } yield key -> idx
+          part.toOption
+        case _ =>
+          None
       }
-    }
+      .groupMap(_._1)(_._2)
 
-    if (metadata.isEmpty) {
-      // empty list
+    if (files.isEmpty) {
+      // no file matched
       false
-    } else if (nums.nonEmpty) {
-      // found xxxxx-of-yyyyy pattern
-      val parts = nums.map(_._1).sorted
-      val total = nums.map(_._2).toSet
-      metadata.size == nums.size && // all paths matched
-      total.size == 1 && total.head == parts.size && // yyyyy part
-      parts.head == 0 && parts.last + 1 == parts.size // xxxxx part
-    } else {
+    } else if (writtenShards.isEmpty) {
+      // assume progress is complete when shard info is not retrieved and files are present
```

Comment:
Is this all we can do here?

Reply:
It was the previous behavior. when user give a custom naming policy/shard template, we do not have this info the the `ReadParameter, thus we can't find shard numbers easily.